### PR TITLE
Feature/validator/discord

### PR DESCRIPTION
**Pull request by fengshanshan**
discord validator,  including
1. code && test logic part of discord validator under the folder "validator"
2. docs for discord validator 
3. fixing test case for keybase
4. optimising output for cli

### CODE REVIEW

File: validator/discord/discord.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+	dg, err := discordgo.New("Bot " + config.C.Platform.Discord.BotToken)
+	if err != nil {
+		return xerrors.Errorf("error creating Discord session, err= %v", err)
+	}
+	rs, err := dg.ChannelMessage(pathArr[3], pathArr[4])
+	if err != nil {
+		return xerrors.Errorf("cannot get the proof err=%v", err)
+	}
+	if strings.ToLower(fmt.Sprintf("%s", rs.Author)) != dc.Identity {
```
Comment (nykma) [inline validator/discord/discord.go:94]: Why? Is `string(rs.Author)` not enough?

File: validator/discord/discord.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+	u, err := url.Parse(dc.ProofLocation)
+	urlPath := path.Clean(u.Path)
+	pathArr := strings.Split(strings.TrimSpace(urlPath), "/")
+	//proof location will be like: https://discord.com/channels/960708146706395176/960708146706395179/961458176719487076
+	if len(pathArr) != 5 {
+		return xerrors.Errorf("haven't got right proof location, err=%v", err)
+	}
+
+	dg, err := discordgo.New("Bot " + config.C.Platform.Discord.BotToken)
```
Comment (nykma) [inline validator/discord/discord.go:86]: Rename `dg` as `client`

File: validator/discord/discord.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+func (dc *Discord) Validate() (err error) {
+	dc.SignaturePayload = dc.GenerateSignPayload()
+
+	u, err := url.Parse(dc.ProofLocation)
+	urlPath := path.Clean(u.Path)
+	pathArr := strings.Split(strings.TrimSpace(urlPath), "/")
+	//proof location will be like: https://discord.com/channels/960708146706395176/960708146706395179/961458176719487076
+	if len(pathArr) != 5 {
+		return xerrors.Errorf("haven't got right proof location, err=%v", err)
```
Comment (nykma) [inline validator/discord/discord.go:83]: Use `%w` to wrap `err` in `xerrors.Errorf()`

No need to add `err=` in Golang best practise.

See my previous `xerrors.Errorf()` for references.

File: validator/discord/discord.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+type Discord struct {
+	*validator.Base
+}
+
+var (
+	re            = regexp.MustCompile(MATCH_TEMPLATE)
+	POST_TEMPLATE = map[string]string{
+		"default": "Verifying my discord ID: %s on NextID. \nSig: %%SIG_BASE64%%",
+		"en-US":   "Verifying my discord ID: %s on NextID. \nSig: %%SIG_BASE64%%",
+		"zh-CN":   "在NextID上认证我的账号:%s。\nSig: %%SIG_BASE64%%",
```
Comment (nykma) [inline validator/discord/discord.go:31]: Space missing before and after `%s`

File: config/main.go [hunk 5]
```diff
@@ -46,6 +47,13 @@ type CliConfig struct {
 	QueryPath  string `json:"query_url"`
 }
 
+type DiscordPlatformConfig struct {
+	ApplicationID        string `json:"application_id"`
+	PublicKey            string `json:"public_key"`
+	BotToken             string `json:"bot_token"`
+	ProofServerChannelID string `json:"proof_server_channel_id"`
+}
```
Comment (nykma) [inline config/main.go:55]: `DiscordPlatformConfig` struct doesn't match `config/config.sample.json` modification. 

Which one is correct?

File: docs/README.md [hunk 6]
```diff
@@ -10,12 +10,13 @@ you on any supported platform.
...
-| Ethereum | `ethereum`       | Wallet address `0x123AbC...` | N/A (Two-way signatures created from persona sk and wallet sk)                 |                                                         |
-| Github   | `github`         | `github_username`            | Public visible Gist ID `a6dddd2811af21b671fd`                                  | Gist should contains `0xPUBKEY_COMRESSED_HEX.json` file |
+| Ethereum | `ethereum`       | Wallet address `0x123AbC...` | N/A (Two-way signatures created from persona sk and wallet sk)                |                                                         |
+| Github   | `github`         | `github_username`            | Public visible Gist ID `a6dddd2811af21b671fd`                                 | Gist should contains `0xPUBKEY_COMRESSED_HEX.json` file |
+| Discord  | `discord`        | `discord_username`           | message link (`https://discord.com/channels/DIGITS/DIGITS/DIGITS`)            |                                                         |
```
Comment (nykma) [inline docs/README.md:19]: `identity` part is misleading. Should be sth. like `DiscordUser#0000`

File: validator/discord/discord.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	//proof location will be like: https://discord.com/channels/960708146706395176/960708146706395179/961458176719487076
+	if len(pathArr) != 5 {
+		return xerrors.Errorf("haven't got right proof location, err=%v", err)
+	}
+
+	dg, err := discordgo.New("Bot " + config.C.Platform.Discord.BotToken)
+	if err != nil {
+		return xerrors.Errorf("error creating Discord session, err= %v", err)
+	}
+	rs, err := dg.ChannelMessage(pathArr[3], pathArr[4])
```
Comment (nykma) [inline validator/discord/discord.go:90]: What is `rs`?

**Review by nykma — CHANGES_REQUESTED** [review]

File: validator/discord/discord.go (hunk 1 above)
Comment (fengshanshan) [reply validator/discord/discord.go:94]: Author is an object. Seems like only in fmt.Sprintf("%s", rs.Author), we can get the username like "discordName#1234", the others, the other field in Author Object, cannot get it.

File: validator/discord/discord.go (hunk 4 above)
Comment (fengshanshan) [reply validator/discord/discord.go:31]: image
still there

**Review by fengshanshan — COMMENTED** [review]
all fixed. replied to some of the comments.

File: validator/discord/discord.go (hunk 4 above)
Comment (nykma) [reply validator/discord/discord.go]: `zh-CN` version.

**Review by fengshanshan — COMMENTED** [review]
fixed.

File: validator/discord/discord.go [hunk 8]
```diff
@@ -78,24 +78,27 @@ func (dc *Discord) Validate() (err error) {
...
 	if err != nil {
-		return xerrors.Errorf("error creating Discord session, err= %v", err)
+		return xerrors.Errorf("Error creating Discord session: %w", err)
 	}
-	rs, err := dg.ChannelMessage(pathArr[3], pathArr[4])
+
+	msgResp, err := client.ChannelMessage(pathArr[3], pathArr[4])
 	if err != nil {
-		return xerrors.Errorf("cannot get the proof err=%v", err)
+		return xerrors.Errorf("Error when get the proof: %w", err)
```
Comment (nykma) [inline validator/discord/discord.go:94]: English grammar

**Review by nykma — APPROVED** [review]

**Pull request merged**